### PR TITLE
Refactor core/hash_map.h

### PR DESCRIPTION
**Pull request by Cradmon**
- Added an underscore to the beginning of private methods.
- Changed `if (!hash_table)` checks to `if (!elements)` to try and be more intuitive.
- Removed `erase_hash_table` and combined its code into `clear`.
- Changed `check_hash_table` to `_update_hash_table` to be more descriptive.
- Changed `create_element` to `_find_or_create_element` and combine the duplicate code of `set` and `operator[]` into it.
- Changed `get_key_value_ptr_array` to return a Pair pointer list, rather than fill one given to it. To try and make the method safer.
- Updated comments.

### CODE REVIEW

File: core/hash_map.h [hunk 1]
```diff
@@ -261,147 +229,97 @@ class HashMap {
...
 
-	/**
-	 * Get a key from data, return a const reference.
-	 * WARNING: this doesn't check errors, use either getptr and check NULL, or check
-	 * first with has(key)
-	 */
-
+	// Get data using a key
+	// WARNING: this doesn't check errors,
+	// use either getptr directly and check NULL, or check first with has(key)
```
Comment (aaronfranke) [inline core/hash_map.h]: Why not leave multi-line comments as multi-line `/* */`? If desired, we could reduce the number of lines by moving the starting and ending ones to the same lines as the text.

Comment (Cradmon) [reply core/hash_map.h]: I'm just used to using single line comments, hence the change. But by trying to mainly use single line comments, it gets me to try and keep comments simple and placed where necessary.

This is a preference of comment syntax. I've changed it to single line comments in this pr but can change it back to multi-line comments, if they are preferred. As long as the syntax remains consistent within the file.

In this case, the first line could be removed entirely, as it just repeats what the function definition already says, so I'll change that.

Comment (aaronfranke) [reply core/hash_map.h]: Agree completely on that last part. When I look at comments I don't want a restatement of the function name (unless the sentence adds more information as well). Caveats and when-to-use/when-not-to-use are great.

**Comment by akien-mga** [comment]
As mentioned in https://github.com/godotengine/godot/pull/22997#issuecomment-494148821, we decided in today's PR review meeting ([logs](http://godot.eska.me/irc-logs/meeting/2019-05-20.log)) not to merge those changes in their current form for the time being. Thanks for the contribution nevertheless!

**Pull request closed without merging**